### DISCUSSION TITLE
LoRaWAN - manually set TX Power level?

### DISCUSSION THREAD

**Discussion by K4KDR** [jgromes/RadioLib#1074] [Q&A] [answered]
Greetings!   Using an example sketch such as **LoRaWAN_ABP**, I am in search of the 'correct' way to lock in the **TX Power** level.

In `src/protocols/LoRaWAN/LoRaWANBands.cpp`, I see that for each bandplan there is a parameter called **.powerMax**.

There is also a specific command parameter (sorry if not the correct term) in
` /src/protocols/LoRaWAN/LoRaWAN.h ` called **setTxPower**

Lastly, associated with each module (sx1276, in my case) there the standard directive **radio.setOutputPower**

What I do NOT know is, which setting gets the "last word" - i.e., has the highest priority.

The context here is 2-fold... for experimentation purposes I would like to be able to guarantee that my device is transmitting at its maximum RF power level (without regard to any setting established by the LoRaWAN bandplan selected)

-and-

I need to be able to impose a much LOWER TX power level to safely use an external RF amp.

Many thanks for any guidance on how to lock in a single, specific RF TX power level using RadioLib's LoRaWAN capabilities.

-Scott,  K4KDR

**Comment by StevenCellist** [comment, marked as answer]
`node.setTxPower()` overrules `radio.setOutputPower()`, but is limited by `band.powerMax`. Make sure that you comply with local regulations & laws (the actual, real laws that may earn you a visit from police) before increasing the `powerMax` value. Otherwise just use `setTxPower()`.

**Comment by K4KDR** [comment]
Thanks a million for the reply - that is very helpful!

And thanks also for the cautionary and very important note about staying within legal power limits.

I'll continue testing now without having to wonder if I am adjusting TX Power in the proper way!

**Reply by StevenCellist** [reply]
If you can `#define RADIOLIB_DEBUG_PROTOCOL 1` in the BuildOptUser.h file, you will get a lot of output from what happens under the hood - you should also be able to see some TX configuration happening when you call setTxPower(). But it may be too overwhelming, as it also spits out lotssss of other details.

**Reply by HeadBoffin** [reply]
But that's because @StevenCellist likes Big Logs, he can not lie.
